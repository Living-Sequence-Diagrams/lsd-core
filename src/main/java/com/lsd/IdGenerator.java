package com.lsd;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;

import static java.util.UUID.randomUUID;

public class IdGenerator {
    private final AtomicInteger counter = new AtomicInteger();

    private final Supplier<String> idSupplier;

    public IdGenerator(boolean isDeterministic) {
        if (isDeterministic) {
            idSupplier = () -> String.valueOf(counter.incrementAndGet());
        } else {
            idSupplier = () -> randomUUID().toString()
                    .replaceAll("-", "");
        }
    }

    public String next() {
        return idSupplier.get();
    }

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    public void reset() {
        counter.set(0);
    }
}
