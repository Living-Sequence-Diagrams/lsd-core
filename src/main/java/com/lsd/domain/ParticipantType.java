package com.lsd.domain;

import org.apache.commons.text.WordUtils;

public enum ParticipantType {
    ACTOR("actor"),
    BOUNDARY("boundary"),
    COLLECTIONS("collections"),
    CONTROL("control"),
    DATABASE("database"),
    ENTITY("entity"),
    PARTICIPANT("participant"),
    QUEUE("queue");

    private final String type;

    ParticipantType(final String type) {
        this.type = type;
    }

    public Participant called(String name) {
        return new Participant(String.format("%s %s", type, removeSpaces(name)));
    }

    public Participant called(String name, String alias) {
        return new Participant(String.format("%s %s as \"%s\"", type, removeSpaces(name), alias));
    }

    private String removeSpaces(String s) {
        return WordUtils.capitalizeFully(s)
                .replaceAll(" ", "");
    }
}
