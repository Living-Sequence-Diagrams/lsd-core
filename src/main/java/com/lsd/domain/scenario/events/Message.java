package com.lsd.domain.scenario.events;

import lombok.Builder;
import lombok.Value;

import static java.lang.String.format;

@Value
@Builder
public class Message implements Interaction {
    String id;
    String label;
    String from;
    String to;
    Object data;

    public String toMarkup() {
        return format("%s -> %s:[[#%s %s]]",
                getFrom(), getTo(), getId(), getLabel());
    }
}
