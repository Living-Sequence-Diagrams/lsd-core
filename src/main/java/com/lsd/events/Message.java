package com.lsd.events;

import com.lsd.report.model.DataHolder;
import lombok.Builder.Default;
import lombok.Data;
import lombok.experimental.SuperBuilder;
import org.apache.commons.text.StringSubstitutor;

import java.util.Map;

import static com.lsd.events.ArrowType.SOLID;

@Data
@SuperBuilder
public class Message implements DataHolder {
    private final String id;
    private final String from;
    private final String to;
    private final String label;

    @Default
    private final Object data = "";

    @Default
    private final String colour = "";

    @Default
    private final ArrowType arrowType = SOLID;

    public String toMarkup() {
        return StringSubstitutor.replace("${from} ${arrow} ${to}: <text fill=\"${colour}\">[[#${id} {${tooltip}} ${label}]]</text>", Map.of(
                "from", getFrom(),
                "to", getTo(),
                "id", getId(),
                "tooltip", getLabel().replaceAll("<\\$.*?>", ""),
                "label", abbreviatedLabel(),
                "colour", getColour(),
                "arrow", getArrowType().toMarkup(getColour())
        ));
    }

}
