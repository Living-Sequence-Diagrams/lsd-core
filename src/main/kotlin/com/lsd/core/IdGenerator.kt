package com.lsd.core

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger

private val randomIdSupplier: () -> String = { randomUUID().toString().replace("-".toRegex(), "") }

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    private val deterministicIdSupplier: () -> String = { counter.incrementAndGet().toString() }

    private val idSupplier: () -> String = if (isDeterministic) deterministicIdSupplier else randomIdSupplier

    fun next(): String = idSupplier.invoke()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}