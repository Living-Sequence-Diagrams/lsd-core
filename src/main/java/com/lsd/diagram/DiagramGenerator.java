package com.lsd.diagram;

import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Participant;
import com.lsd.report.model.SequenceDiagram;
import com.lsd.events.SequenceEvent;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.toList;

@Builder
public class DiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<SequenceEvent> events;

    public SequenceDiagram sequenceDiagram() {
        String uml = generateSequenceUml();
        String svg = generateSequenceSvg(uml);
        return SequenceDiagram.builder()
                .uml(uml)
                .svg(svg)
                .build();
    }

    private String generateSequenceSvg(String markup) {
        SvgConverter svgConverter = new SvgConverter();

        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateSequenceUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants,
                "events", events.stream()
                        .map(SequenceEvent::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
