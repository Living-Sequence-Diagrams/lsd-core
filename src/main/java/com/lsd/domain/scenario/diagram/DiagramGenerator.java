package com.lsd.domain.scenario.diagram;

import com.lsd.domain.Participant;
import com.lsd.domain.scenario.events.Event;
import com.lsd.domain.scenario.events.Interaction;
import com.lsd.properties.LsdProperties;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.toList;

@Builder
public class DiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<Event> events;

    public SequenceDiagram sequenceDiagram() {
        return SequenceDiagram.builder()
                .uml(generateSequenceUml())
                .interactions(events.stream()
                        .filter(Interaction.class::isInstance)
                        .map(Interaction.class::cast)
                        .collect(toList()))
                .build();
    }

    @SneakyThrows
    private String generateSequenceUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants,
                "events", events.stream()
                        .map(Event::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
