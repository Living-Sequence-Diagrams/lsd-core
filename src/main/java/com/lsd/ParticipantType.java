package com.lsd;

import com.lsd.report.model.Participant;
import com.lsd.diagram.ValidComponentName;

public enum ParticipantType {
    ACTOR("actor"),
    BOUNDARY("boundary"),
    COLLECTIONS("collections"),
    CONTROL("control"),
    DATABASE("database"),
    ENTITY("entity"),
    PARTICIPANT("participant"),
    QUEUE("queue");

    private final String type;

    ParticipantType(final String type) {
        this.type = type;
    }

    public Participant called(String name) {
        return new Participant(String.format("%s %s", type, ValidComponentName.of(name)));
    }

    public Participant called(String name, String alias) {
        return new Participant(String.format("%s %s as \"%s\"", type, ValidComponentName.of(name), alias));
    }
}
