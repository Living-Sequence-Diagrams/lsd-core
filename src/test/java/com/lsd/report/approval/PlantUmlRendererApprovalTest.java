package com.lsd.report.approval;

import com.lsd.IdGenerator;
import com.lsd.ParticipantType;
import com.lsd.diagram.DiagramGenerator;
import com.lsd.events.Message;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.LinkedHashSet;
import java.util.List;

import static com.lsd.events.ArrowType.BI_DIRECTIONAL_DOTTED;
import static com.lsd.events.ArrowType.DOTTED;

class PlantUmlRendererApprovalTest {

    @Test
    void renderSequenceUml() {
        Approvals.verify(diagramGenerator().sequenceDiagram().getUml());
    }

    private DiagramGenerator diagramGenerator() {
        return DiagramGenerator.builder()
                .idGenerator(new IdGenerator(true))
                .includes(new LinkedHashSet<>(List.of(
                        "tupadr3/font-awesome-5/clock",
                        "tupadr3/font-awesome-5/database"
                )))
                .participants(List.of(
                        ParticipantType.ACTOR.called("Arnold", "Arnie"),
                        ParticipantType.BOUNDARY.called("Nick"),
                        ParticipantType.CONTROL.called("A control"),
                        ParticipantType.COLLECTIONS.called("some collection"),
                        ParticipantType.ENTITY.called("Entity"),
                        ParticipantType.PARTICIPANT.called("Party"),
                        ParticipantType.QUEUE.called("Party"),
                        ParticipantType.DATABASE.called("Derek", "D for danger")
                ))
                .events(List.of(
                        Message.builder()
                                .id("1")
                                .label("Sending a request")
                                .from("A")
                                .to("B")
                                .data("Please help?")
                                .build(),
                        Message.builder()
                                .id("2")
                                .label("Sending a response with a long label that should be abbreviated if all goes well")
                                .from("B")
                                .to("A")
                                .data("OK")
                                .arrowType(DOTTED)
                                .build(),
                        Message.builder()
                                .id("3")
                                .label("On the phone")
                                .from("B")
                                .to("A")
                                .arrowType(BI_DIRECTIONAL_DOTTED)
                                .build()
                )).build();
    }
}