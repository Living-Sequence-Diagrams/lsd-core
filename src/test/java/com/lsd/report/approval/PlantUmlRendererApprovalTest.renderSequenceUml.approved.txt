@startuml

!theme plain
skinparam backgroundColor transparent

hide unlinked

!include <tupadr3/font-awesome-5/clock>
!include <tupadr3/font-awesome-5/database>

actor Arnold as "Arnie"
boundary Nick
control AControl
collections SomeCollection
entity Entity
participant Party
queue Party
database Derek as "D for danger"

A -> B: <text fill="">[[#1 {Sending a request} Sending a request]]</text>
B --> A: <text fill="">[[#2 {Sending a response with a long label that should be abbreviated if all goes well} Sending a response with a long label that shoul...]]</text>
B <--> A: <text fill="">[[#3 {On the phone} On the phone]]</text>

@enduml